pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,505 @@
+
+				Ok(())
+			}
+			// Proof entries
+			(InboundEntry::Proof(self_proof_entry), InboundEntry::Proof(other_proof_entry)) => {
+				if self_proof_entry.session_id != other_proof_entry.session_id {
+					*self = other;
+
+					return Ok(());
+				}
+
+				self_proof_entry
+					.current_count
+					.ensure_add_assign(other_proof_entry.current_count)?;
+
+				Ok(())
+			}
+			// Mismatches
+			(InboundEntry::Message(_), InboundEntry::Proof(_)) => {
+				Err(Error::<T>::ExpectedMessageType.into())
+			}
+			(InboundEntry::Proof(_), InboundEntry::Message(_)) => {
+				Err(Error::<T>::ExpectedMessageProofType.into())
+			}
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Retrieves all stored routers and then filters them based
+	/// on the available routers for the provided domain.
+	pub(crate) fn get_router_ids_for_domain(
+		domain: Domain,
+	) -> Result<Vec<T::RouterId>, DispatchError> {
+		let stored_routers = Routers::<T>::get();
+
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let res = stored_routers
+			.iter()
+			.filter(|stored_router| {
+				all_routers_for_domain
+					.iter()
+					.any(|available_router| *stored_router == available_router)
+			})
+			.cloned()
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	pub(crate) fn get_expected_proof_count(
+		router_ids: &[T::RouterId],
+	) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count.saturated_into())
+	}
+
+	/// Gets the message proof for a message.
+	pub(crate) fn get_message_proof(message: T::Message) -> Proof {
+		match message.get_proof() {
+			None => message
+				.to_proof_message()
+				.get_proof()
+				.expect("message proof ensured by 'to_message_proof'"),
+			Some(proof) => proof,
+		}
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	pub(crate) fn upsert_pending_entry(
+		message_proof: Proof,
+		router_id: &T::RouterId,
+		new_inbound_entry: InboundEntry<T>,
+	) -> DispatchResult {
+		PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+			match storage_entry {
+				None => {
+					*storage_entry = Some(new_inbound_entry);
+
+					Ok::<(), DispatchError>(())
+				}
+				Some(stored_inbound_entry) => {
+					stored_inbound_entry.pre_dispatch_update(new_inbound_entry)
+				}
+			}
+		})
+	}
+
+	/// Checks if the number of proofs required for executing one message
+	/// were received, and if so, decreases the counts accordingly and executes
+	/// the message.
+	pub(crate) fn execute_if_requirements_are_met(
+		message_proof: Proof,
+		router_ids: &[T::RouterId],
+		session_id: T::SessionId,
+		expected_proof_count: u32,
```

Comment:
These 3 can be obtained from the domain to avoid passing them, right? No strong opinion anyways

Reply:
Yes, but we also want to do the minimum amount of reading from storage. I can pass the domain here and retrieve the router IDs again but that will feel off no?

Reply:
Mmm... yeah, no strong opinion. Makes sense what you said 👍🏻.

Just as a note, reading multiple times the same storage only increments the reads by 1.